packages/core/src/server.ts

```diff
@@ -581,6 +582,12 @@ export class Controller extends API {
         // authentication.
         if (!(await getCryptoProvider().timingSafeEqual(M, srp.M1!))) {
             this.log("account.createSession", { success: false });
+            if (acc.settings.failedLoginAttemptNotifications) {
+                try {
+                    // Send invite link to invitees email address
+                    await this.messenger.send(acc.email, new FailedLoginAttemptMessage({ srpId }));
+                } catch (e) {}
+            }
```

Comment:
Hm, maybe we should add a number of allowed retries? I.e. only send a notification after x failed login attempts?

Reply:
Well, then we need to store that. We do it for the auth request, but at this stage you _know_ something or _have access_ to something (like the email, to get the code). I think it's safe to just send the email here because of it.

Reply:
We could simply store the number of attempts in the `SRPSession` object, which would be trivial. My concern is that people will often mistype their password at least once or twice before getting it right (I know I do) and then those emails will get annoying very quickly, causing them to disable them. On the other hand an attacker who doesn't get the password right on the first try will probably not get it right on the second or third try, either. So I think it's safe to give them a handful of tries, at least.

Reply:
Actually how about this: Give the user a maximum of `x` tries to get the password right (~5 should be a good number, but we can make that configurable). Once they exceed that number of tries, cancel the SRP session, remove the device from trusted devices and send the email notification. Then the notification could say something along the lines of "We've prevented a login attempt from ..." and then the device model and location.

Reply:
The device model and location can be found in `session.deviceInfo`.